corehq/apps/hqcase/api/updates.py

```diff
@@ -73,11 +76,23 @@ def _if_specified(value):
             update=dict(self.properties),
             close=self.close,
             index={
-                name: IndexAttrs(index.case_type, index.case_id, index.relationship)
-                for name, index in self.indices.items()
+                name: IndexAttrs(
+                    case_type=index.case_type,
+                    case_id=self._get_index_id(index, case_db),
+                    relationship=index.relationship
+                ) for name, index in self.indices.items()
             },
         ).as_text()
 
+    def _get_index_id(self, index, case_db):
+        if index.case_id:
+            return index.case_id
+        if index.temporary_id:
+            return case_db.lookup_id(index.temporary_id, 'temporary_id')
+        if index.external_id:
+            return case_db.lookup_id(index.external_id, 'external_id')
```

Comment:
nit: what about specific functions for these rather than passing in strings? I think it would reduce the likelihood of typos and makes it easier to know what the available options are.
```suggestion
            return case_db.lookup_by_external_id(index.external_id)
```

Reply:
Thanks yeah - and storing both in a single dict didn't really make sense anymore with the way I'd restructured this.  Pushing an update shortly.